#67Added Binding Properties  Pattern (#1776)
iluwatar/java-design-patterns: binding-properties/README.md

  }
  public abstract void setValue(T newValue);
  public abstract T getValue();
}
```
In real-world examples, binding properties pattern is used along with event handlers. Here we use an implementation of functional interface for calculating the new value to substitute the event handler. Whenever the set method is invoked, it should notify all the binded properties to change according to a giving function. We then add unidirectional and bidirectional bind methods for it. The default changing behaviour is to let a pair of binded properties to keep mutually equal.</br>

What if we want to define some behaviors other than keeping the properties same? We can add the following code into the abstract class:</br>
```java
public <M> void bind(BindableField<M> target, Function<T, M> onChangeFunction) {
  notifyMap.put(target, onChangeFunction);
}

public <M> void bidirectionalBind(BindableField<M> target,
                                  Function<T, M> onChangeFunction,
                                  Function<M, T> oppositeOnChangeFunction) {
  notifyMap.put(target, onChangeFunction);
  target.bind(this, oppositeOnChangeFunction);
}
```

> Becareful, circulate bindings might introduce infinity loops. Here we can allow the property to resist changing temporarily:</br>

Reviewer: ```suggestion
Be careful, circulate bindings might introduce infinite loops. Here we can allow the property to resist changing temporarily:</br>

```

```java
protected boolean blockPropagation;
protected void block() {
  blockPropagation = true;
}

protected void unBlock() {
  blockPropagation = false;
}
```

> If the propagation is block, the set method invocation should be ignored.</br>

Reviewer: ```suggestion
If the propagation is blocked, the set method invocation should be ignored.</br>

```

Inherit the class to build concrete bindable properties:</br>
```java
class BindableDouble extends BindableField<Double> {

  public BindableDouble(Double value) {
    this.value = value;
    notifyMap = new HashMap<>();
  }

  @Override
  public void setValue(Double newValue) {
    if (blockPropagation) {
      return;
    }
    beforePropertyChanged(newValue);
    value = newValue;
  }

  @Override
  public Double getValue() {
    return value;
  }
}
public class BindableInteger extends BindableField<Integer> {

  public BindableInteger(Integer value) {
    this.value = value;
    notifyMap = new HashMap<>();
  }

  @Override
  public void setValue(Integer newValue) {
    if (blockPropagation) {
      return;
    }
    beforePropertyChanged(newValue);
    value = newValue;
  }

  @Override
  public Integer getValue() {
    return value;
  }
}
```

> Now we can make use of our binding properties. We have a little warrior that has HP, MP (for casting spells) and some remaining lives. Every time the little warrior's HP is healed or damaged, he will gain the same amount of recovering or damaging on MP. Whenever the HP is lower than 0, he must respawn himself until he does not have remaining lives. Besides, we need to display his HP and MP percentages on the screen. Now we need a mechanism to let all the values changes simultaneously. So we let his HP, MP, HP percentage, MP percentage to be bindable:</br>

Reviewer: ```suggestion
Now we can make use of our binding properties. We have a little warrior that has HP, MP (for casting spells), and some remaining lives. Every time the little warrior's HP is healed or damaged, he will gain the same amount of recovering or damaging on MP. Whenever the HP is lower than 0, he must respawn himself until he does not have remaining lives. Besides, we need to display his HP and MP percentages on the screen. Now we need a mechanism to let all the values change simultaneously. So we let his HP, MP, HP percentage, MP percentage be bindable:</br>

```

            (newPercentage) -> Math.min(Math.max(newPercentage / 100.0 * maxHp, 0.0), 50.0));
    myMp.bidirectionalBind(myMpPercentage,
            (newMp) -> Math.min(Math.max((newMp / maxMp) * 100.0, 0.0), 100.0),
            (newPercentage) -> Math.min(Math.max(newPercentage / 100.0 * maxMp, 0.0), 50.0));
    myHp.bind(remainLives,
            newHp -> newHp <= 0.0 ? remainLives.getValue() - 1 : remainLives.getValue());
  }

  public void displayMyStatus() {
    LOGGER.info("HP:{}  MP:{}  HP percentage:{}  MP percentage:{}  Remaining Lives:{}",
            myHp.getValue(), myMp.getValue(),
            myHpPercentage.getValue(), myMpPercentage.getValue(), remainLives.getValue());
  }

  public void damage(double value) {
    myHp.setValue(Math.min(Math.max(myHp.getValue() - value, 0), maxHp));
  }

  public void heal(double value) {
    myHp.setValue(Math.min(Math.max(myHp.getValue() + value, 0), maxHp));
  }

  public void castSpell(double cost) {
    myMp.setValue(Math.min(Math.max(myMp.getValue() - cost, 0), maxMp));
  }

  public int getRemainLives() {
    return remainLives.getValue();
  }
}
```

> Now we can let his advanture begin to check whether the data changing dependencies are correctly built:</br>

Reviewer: ```suggestion
Now we can let his adventure begin to check whether the data changing dependencies are correctly built:</br>

```

---
layout: pattern

> title: Binding Properties Pattern

Reviewer: ```suggestion
title: Binding Properties

```

folder: binding-properties

> permalink: /patterns/query-object/

Reviewer: ```suggestion
permalink: /patterns/binding-properties/

```
